packages/flutter/lib/src/services/raw_keyboard_macos.dart

```diff
@@ -0,0 +1,315 @@
+/// `NSEvent` interface.
+///
+/// See also:
+///
+///  * [RawKeyboard], which uses this interface to expose key data.
+class RawKeyEventDataMacOs extends RawKeyEventData {
+  /// Creates a key event data structure specific for macOS.
+  ///
+  /// The [characters], [charactersIgnoringModifiers], and [modifiers], arguments
+  /// must not be null.
+  const RawKeyEventDataMacOs(
+      {this.characters = '',
+      this.charactersIgnoringModifiers = '',
+      this.keyCode = 0,
+      this.modifiers = 0})
+      : assert(characters != null),
+        assert(charactersIgnoringModifiers != null),
+        assert(keyCode != null),
+        assert(modifiers != null);
+
+  /// The Unicode characters associated with a key-up or key-down event.
+  ///
+  /// See also:
+  ///
+  ///   * [Apple's NSEvent documentation](https://developer.apple.com/documentation/appkit/nsevent/1534183-characters?language=objc)
+  final String characters;
+
+  /// The characters generated by a key event as if no modifier key (except for
+  /// Shift) applies.
+  ///
+  /// See also:
+  ///
+  ///   * [Apple's NSEvent documentation](https://developer.apple.com/documentation/appkit/nsevent/1524605-charactersignoringmodifiers?language=objc)
+  final String charactersIgnoringModifiers;
+
+  /// The virtual key code for the keyboard key associated with a key event.
+  ///
+  /// See also:
+  ///
+  ///   * [Apple's NSEvent documentation](https://developer.apple.com/documentation/appkit/nsevent/1534513-keycode?language=objc)
+  final int keyCode;
+
+  /// A mask of the current modifiers using the values in Modifier Flags.
+  ///
+  /// See also:
+  ///
+  ///   * [Apple's NSEvent documentation](https://developer.apple.com/documentation/appkit/nsevent/1535211-modifierflags?language=objc)
+  final int modifiers;
+
+  @override
+  String get keyLabel => charactersIgnoringModifiers.isEmpty ? null : charactersIgnoringModifiers;
+
+  @override
+  PhysicalKeyboardKey get physicalKey => kMacOsToPhysicalKey[keyCode] ?? PhysicalKeyboardKey.none;
+
+  @override
+  LogicalKeyboardKey get logicalKey {
+    // Look to see if the keyCode is a printable number pad key, so that a
+    // difference between regular keys (e.g. "=") and the number pad version
+    // (e.g. the "=" on the number pad) can be determined.
+    final LogicalKeyboardKey numPadKey = kMacOsNumPadMap[keyCode];
+    if (numPadKey != null) {
+      return numPadKey;
+    }
+
+    // Look to see if the keyCode is one we know about and have a mapping for.
+    if (keyLabel != null &&
+        !LogicalKeyboardKey.isControlCharacter(keyLabel)) {
+      assert(charactersIgnoringModifiers.length <= 2);
+      int codeUnit = charactersIgnoringModifiers.codeUnitAt(0);
+      if (charactersIgnoringModifiers.length == 2) {
+        // Not covering length > 2 case since > 1 is already unlikely.
+        final int secondCode = charactersIgnoringModifiers.codeUnitAt(1);
+        codeUnit |= secondCode;
```

Comment:
You're going to want to shift one of the codes by 16 bits.

Probably it should be this:
```suggestion
        codeUnit = (codeUnit << 16) | secondCode;
```